src/main/scala/is/hail/io/index/IndexReader.scala

```diff
@@ -83,25 +84,154 @@ class IndexReader(hConf: Configuration, path: String, cacheCapacity: Int = 8) ex
     }
   }
 
-  private def queryByIndex(index: Long, level: Int, offset: Long): LeafChild = {
+  // Returns smallest i, 0 <= i < n, for which p(i) holds, or returns n if p(i) is false for all i.
+  // Assumes all i for which p(i) is true are greater than all i for which p(i) is false.
+  // Returned value j has the property that p(i) is false for all i in [0, j),
+  // and p(i) is true for all i in [j, n).
+  private def findPartitionPoint(n: Int, p: (Int) => Boolean): Int = {
+    var left = 0
+    var right = n
+    while (left < right) {
+      val mid = left + (right - left) / 2
+      if (p(mid))
+        right = mid
+      else
+        left = mid + 1
+    }
+    left
+  }
+
+  // Returns smallest i, 0 <= i < n, for which f(i) >= key, or returns n if f(i) < key for all i
+  private[io] def binarySearchLowerBound(n: Int, key: Annotation, f: (Int) => Annotation): Int =
+    findPartitionPoint(n, i => ordering.gteq(f(i), key))
+
+  // Returns smallest i, 0 <= i < n, for which f(i) > key, or returns n if f(i) > key for all i
+  private[io] def binarySearchUpperBound(n: Int, key: Annotation, f: (Int) => Annotation): Int =
+    findPartitionPoint(n, i => ordering.gt(f(i), key))
+
+  private def lowerBound(key: Annotation, level: Int, offset: Long): Long = {
+    if (level == 0) {
+      val node = readLeafNode(offset)
+      val idx = binarySearchLowerBound(node.children.length, key, { i => node.children(i).key })
+      node.firstIndex + idx
+    } else {
+      val node = readInternalNode(offset)
+      val children = node.children
+      val n = children.length
+      if (n == 0)
+        return node.firstIndex
+      val idx = binarySearchLowerBound(n, key, { i => children(i).lastKey })
```

Comment:
I think I see how to get rid of `lastKey` on `InternalChild`.

First, in addition to assuming the node at `offset` is nonEmpty, let's add another invariant to `lowerBound`: that `key > node.children.head.firstKey` (or `key > node.children.head.key` if we're at a leaf node). We can enforce this at the start of the recursion by adding a check to the non-recursive method
```scala
private[io] def lowerBound(key: Annotation): Long = {
  if (nKeys == 0 || ordering.lteq(key, readInternalNode(metadata.rootOffset).children.head.firstKey))
    0
  else
    lowerBound(key, height - 1, metadata.rootOffset)
}
```
This is correct because if `key <= root.children.head.firstKey`, then `key` is <= the key at index 0, so the lower bound is 0.

Making use of the stronger invariant, we can rewrite the last two lines of the recursive method:
```scala
val idx = binarySearchLowerBound(n, key, { i => children(i).firstKey })
lowerBound(key, level - 1, children(idx - 1).indexFileOffset)
```
By the definition of lower bound, `idx` is the smallest `i` such that `children(i).firstKey >= key`. But by the invariant, `children(0).firstKey < key`, so `idx > 0`. Therefore, `idx - 1` is a valid index, and is the largest `i` such that `children(i).firstKey < key`. That shows that the recursive call maintains the invariant. As for correctness, clearly `children(idx-1).firstIndex < lowerBound <= children(i).firstIndex`. The subtlest case is if all keys in the subtree rooted at `children(idx-1)` are `> key`. But then we will end up at the right-most leaf node `leaf` in the subtree, and the `binarySearchLowerBound` call will return `leaf.children.length`, so `lowerBound` will return `leaf.firstIndex + leaf.children.length`, which is `children(i).firstIndex`.

`upperBound` can be modified similarly, adding the invariant `key >= node.children.head.firstKey`. Then the root call becomes
```scala
private[io] def upperBound(key: Annotation): Long = {
  if (nKeys == 0 || ordering.lt(key, readInternalNode(metadata.rootOffset).children.head.firstKey))
    0
  else
    lowerBound(key, height - 1, metadata.rootOffset)
}
```
and the last few lines of the recursive method become
```scala
val idx = binarySearchUpperBound(n, key, { i => children(i).firstKey })
upperBound(key, level - 1, children(idx - 1).indexFileOffset)
```
By the definition of upper bound, `idx` is the smallest `i` such that `children(i).firstKey > key`. But by the invariant, `children(0).firstKey <= key`, so `idx > 0`. Therefore, `idx - 1` is a valid index, and is the largest `i` such that `children(i).firstKey <= key`. Just like before, we have `children(idx-1).firstIndex < upperBound <= children(i).firstIndex`.